homeassistant/components/plex/__init__.py

```diff
@@ -1 +1,195 @@
+
+from .const import (
+    CONF_USE_EPISODE_ART,
+    CONF_SHOW_ALL_CONTROLS,
+    CONF_REMOVE_UNAVAILABLE_CLIENTS,
+    CONF_CLIENT_REMOVE_INTERVAL,
+    DEFAULT_HOST,
+    DEFAULT_PORT,
+    DEFAULT_SSL,
+    DEFAULT_VERIFY_SSL,
+    DOMAIN as PLEX_DOMAIN,
+    PLATFORMS,
+    PLEX_CONFIG_FILE,
+    PLEX_MEDIA_PLAYER_OPTIONS,
+    SERVERS,
+)
+from .server import PlexServer
+
+MEDIA_PLAYER_SCHEMA = vol.Schema(
+    {
+        vol.Optional(CONF_USE_EPISODE_ART, default=False): cv.boolean,
+        vol.Optional(CONF_SHOW_ALL_CONTROLS, default=False): cv.boolean,
+        vol.Optional(CONF_REMOVE_UNAVAILABLE_CLIENTS, default=True): cv.boolean,
+        vol.Optional(CONF_CLIENT_REMOVE_INTERVAL, default=600): cv.positive_int,
+    }
+)
+
+SERVER_CONFIG_SCHEMA = vol.Schema(
+    {
+        vol.Optional(CONF_HOST, default=DEFAULT_HOST): cv.string,
+        vol.Optional(CONF_TOKEN): cv.string,
+        vol.Optional(CONF_PORT, default=DEFAULT_PORT): cv.port,
+        vol.Optional(CONF_SSL, default=DEFAULT_SSL): cv.boolean,
+        vol.Optional(CONF_VERIFY_SSL, default=DEFAULT_VERIFY_SSL): cv.boolean,
+        vol.Optional(MP_DOMAIN, default={}): MEDIA_PLAYER_SCHEMA,
+    }
+)
+
+CONFIG_SCHEMA = vol.Schema({PLEX_DOMAIN: SERVER_CONFIG_SCHEMA}, extra=vol.ALLOW_EXTRA)
+
+CONFIGURING = "configuring"
+_LOGGER = logging.getLogger(__package__)
+
+
+def setup(hass, config):
+    """Set up the Plex component."""
+
+    def server_discovered(service, info):
+        """Pass back discovered Plex server details."""
+        if hass.data[PLEX_DOMAIN][SERVERS]:
+            _LOGGER.debug("Plex server already configured, ignoring discovery.")
+            return
+        _LOGGER.debug("Discovered Plex server: %s:%s", info["host"], info["port"])
+        setup_plex(discovery_info=info)
+
+    def setup_plex(config=None, discovery_info=None, configurator_info=None):
+        """Return assembled server_config dict."""
+        json_file = hass.config.path(PLEX_CONFIG_FILE)
+        file_config = load_json(json_file)
+
+        if config:
+            server_config = config
+            host_and_port = (
+                f"{server_config.pop(CONF_HOST)}:{server_config.pop(CONF_PORT)}"
+            )
+            if MP_DOMAIN in server_config:
+                hass.data[PLEX_MEDIA_PLAYER_OPTIONS] = server_config.pop(MP_DOMAIN)
+        elif file_config:
+            _LOGGER.debug("Loading config from %s", json_file)
+            host_and_port, server_config = file_config.popitem()
+            server_config[CONF_VERIFY_SSL] = server_config.pop("verify")
+        elif discovery_info:
+            server_config = {}
+            host_and_port = f"{discovery_info[CONF_HOST]}:{discovery_info[CONF_PORT]}"
+        elif configurator_info:
+            server_config = configurator_info
+            host_and_port = server_config["host_and_port"]
+        else:
+            discovery.listen(hass, SERVICE_PLEX, server_discovered)
```

Comment:
I think we should register only once during component setup, regardless of case. Why did we move this?

Reply:
If setup is called via discovery with no config for the component and no plex.conf file, it would likely fail in setup_plex and fail to load the component. This ensures discovery is set up with an empty config and the component reports a proper load.

Reply:
Is it correct to not register the discovery listener if we load via config?

Reply:
Only one server is allowed, so yes.